Use NativePromise for VideoDecoder and VideoEncoder create methods

youennf opened the pull request:
#### 9b9b91b47d7e40b5a5a4e75c7af74507731a3566
<pre>
Use NativePromise for VideoDecoder and VideoEncoder create methods
<a href="https://rdar.apple.com/137864703">rdar://137864703</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=281415">https://bugs.webkit.org/show_bug.cgi?id=281415</a>

Reviewed by Jean-Yves Avenard.

Make VideoDecoder::create and VideoEncoder::create return promises instead of taking a create callback.
Update the client code accordingly, which makes code more readable.

This notably simplifies the isConfigSupported methods since we can use ScriptExecutionContext::enqueueTaskWhenSettled,
which allows to capture the DeferredPromise object in the native promise handler.

* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::configure):
(WebCore::WebCodecsVideoDecoder::isConfigSupported):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
(WebCore::WebCodecsVideoEncoder::configure):
(WebCore::WebCodecsVideoEncoder::isConfigSupported):
* Source/WebCore/platform/VideoDecoder.cpp:
(WebCore::VideoDecoder::create):
* Source/WebCore/platform/VideoDecoder.h:
* Source/WebCore/platform/VideoEncoder.cpp:
(WebCore::VideoEncoder::create):
* Source/WebCore/platform/VideoEncoder.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
(WebCore::MediaRecorderPrivateWriterWebM::appendVideoFrame):

Canonical link: <a href="https://commits.webkit.org/285184@main">https://commits.webkit.org/285184@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58d7f445d169e9b619fb1bb62a63b8874f84ca2d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75764 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22855 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22675 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56590 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15071 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74716 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46325 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61717 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37039 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42986 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19184 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21196 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64890 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19547 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77484 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15884 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18730 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64306 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15927 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61750 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64303 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15867 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12451 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6089 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46863 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1642 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47934 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49218 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47676 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
```diff
@@ -139,17 +145,7 @@ ExceptionOr<void> WebCodecsVideoDecoder::configure(ScriptExecutionContext& conte
-                decoder.processControlMessageQueue();
-            });
-        }, [identifier, weakThis = ThreadSafeWeakPtr { *this }, decoderCount = ++m_decoderCount](auto&& result) {
+        auto createDecoderPromise = VideoDecoder::create(config.codec, createVideoDecoderConfig(config), [identifier, weakThis = ThreadSafeWeakPtr { *this }, decoderCount = ++m_decoderCount](auto&& result) {
```
```suggestion
        Ref createDecoderPromise = VideoDecoder::create(config.codec, createVideoDecoderConfig(config), [identifier, weakThis = ThreadSafeWeakPtr { *this }, decoderCount = ++m_decoderCount](auto&& result) {
```

youennf replied:
Done

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
```diff
@@ -171,6 +167,19 @@ ExceptionOr<void> WebCodecsVideoDecoder::configure(ScriptExecutionContext& conte
+            auto protectedThis = weakThis.get();
+            if (!protectedThis)
+                return;
+            if (!result.has_value()) {
```
```suggestion
            if (!result) {
```

youennf replied:
Done

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
```diff
@@ -171,6 +167,19 @@ ExceptionOr<void> WebCodecsVideoDecoder::configure(ScriptExecutionContext& conte
+                protectedThis->closeDecoder(Exception { ExceptionCode::NotSupportedError, WTFMove(result.error()) });
+                return;
+            }
+            protectedThis->setInternalDecoder(WTFMove(result.value()));
```
```suggestion
            protectedThis->setInternalDecoder(WTFMove(*result));
```

youennf replied:
Done

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
```diff
@@ -246,20 +255,15 @@ void WebCodecsVideoDecoder::isConfigSupported(ScriptExecutionContext& context, W
-            }
-        });
-    }, [](auto&&) {
+    Ref createDecoderPromise = VideoDecoder::create(config.codec, createVideoDecoderConfig(config), [](auto&&) { });
```
```suggestion
    Ref createDecoderPromise = VideoDecoder::create(config.codec, createVideoDecoderConfig(config), [] { });
```

youennf replied:
Does not work since this is not a promise callback.

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
```diff
@@ -246,20 +255,15 @@ void WebCodecsVideoDecoder::isConfigSupported(ScriptExecutionContext& context, W
+    Ref createDecoderPromise = VideoDecoder::create(config.codec, createVideoDecoderConfig(config), [](auto&&) { });
+
+    context.enqueueTaskWhenSettled(WTFMove(createDecoderPromise), TaskSource::MediaElement, [config, description = WTFMove(description), promise = WTFMove(promise)](auto&& result) mutable {
+        bool success = result.has_value();
```
```suggestion
        bool success = !!result;
```

youennf replied:
Done

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
```diff
@@ -227,6 +222,20 @@ ExceptionOr<void> WebCodecsVideoEncoder::configure(ScriptExecutionContext& conte
+            auto protectedThis = weakThis.get();
+            if (!protectedThis)
+                return;
+            if (!result.has_value()) {
```
```suggestion
           if (!result) {
```

youennf replied:
Done

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
```diff
@@ -227,6 +222,20 @@ ExceptionOr<void> WebCodecsVideoEncoder::configure(ScriptExecutionContext& conte
+                protectedThis->closeEncoder(Exception { ExceptionCode::NotSupportedError, WTFMove(result.error()) });
+                return;
+            }
+            protectedThis->setInternalEncoder(WTFMove(result.value()));
```
```suggestion
            protectedThis->setInternalEncoder(WTFMove(*result));
```

youennf replied:
Done

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
```diff
@@ -344,16 +353,10 @@ void WebCodecsVideoEncoder::isConfigSupported(ScriptExecutionContext& context, W
-        });
-    }, [](auto&&) {
-    }, [](auto&&) {
+    Ref createEncoderPromise = VideoEncoder::create(config.codec, encoderConfig.releaseReturnValue(), [](auto&&) { }, [](auto&&) { });
```
`(auto&&)` is optional with NativePromise handler; don't need it

youennf replied:
But it is not a NativePromise handler, it is a function that can get called multiple times.

jyavenard reviewed (approved):
LGTM, some nits

jyavenard commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -294,6 +285,16 @@ static String codecStringForMediaVideoCodecId(FourCharCode codec)
+                return;
+            m_videoEncoder = result.value().moveToUniquePtr();
+            m_videoEncoder->setRates(m_videoBitsPerSecond, 0);
+            encodePendingVideoFrames();
```
the code was synchronous before as the callback was immediately Called; likely the reason for the current peer-connection webm recorder error.

I think the fix is to call `maybeStartWriting()` here. The first frame will be dropped for now, but I'm fixing this in #35136

youennf replied:
OK, will update PR accordingly.
Maybe we should rename maybeStartWriting() since it is not really starting to write.

youennf committed:
Use NativePromise for VideoDecoder and VideoEncoder create methods
rdar://137864703
https://bugs.webkit.org/show_bug.cgi?id=281415

Reviewed by Jean-Yves Avenard.

Make VideoDecoder::create and VideoEncoder::create return promises instead of taking a create callback.
Update the client code accordingly, which makes code more readable.

This notably simplifies the isConfigSupported methods since we can use ScriptExecutionContext::enqueueTaskWhenSettled,
which allows to capture the DeferredPromise object in the native promise handler.

* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::configure):
(WebCore::WebCodecsVideoDecoder::isConfigSupported):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
(WebCore::WebCodecsVideoEncoder::configure):
(WebCore::WebCodecsVideoEncoder::isConfigSupported):
* Source/WebCore/platform/VideoDecoder.cpp:
(WebCore::VideoDecoder::create):
* Source/WebCore/platform/VideoDecoder.h:
* Source/WebCore/platform/VideoEncoder.cpp:
(WebCore::VideoEncoder::create):
* Source/WebCore/platform/VideoEncoder.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
(WebCore::MediaRecorderPrivateWriterWebM::appendVideoFrame):

Canonical link: https://commits.webkit.org/285184@main
